docker_test.go

```diff
@@ -261,198 +355,63 @@ func run(m *testing.M) (int, error) {
-		},
-		Env: []string{"ZOOKEEPER_CLIENT_PORT=2181"},
-	})
-	if err != nil {
-		log.Panic(err)
-	}
+	SetZookeeper()
 	defer func() {
 		if err := pool.Purge(z); err != nil {
 			log.Printf("Could not purge resource: %s \n", err)
 		}
 	}()
-
-	// Set Kafka: pulls an image, creates a container based on it and runs it
-	KAFKA_ZOOKEEPER_CONNECT := fmt.Sprintf("KAFKA_ZOOKEEPER_CONNECT= zookeeper:%s", z.GetPort("2181/tcp"))
-	log.Println("KAFKA_ZOOKEEPER_CONNECT:", KAFKA_ZOOKEEPER_CONNECT)
-
-	brokerPortInt, err := freeport.GetFreePort()
-	if err != nil {
-		log.Panic(err)
-	}
-	brokerPort = fmt.Sprintf("%s/tcp", strconv.Itoa(brokerPortInt))
-	log.Println("broker Port:", brokerPort)
-
-	localhostPortInt, err = freeport.GetFreePort()
-	if err != nil {
-		log.Panic(err)
-	}
-	localhostPort = fmt.Sprintf("%s/tcp", strconv.Itoa(localhostPortInt))
-	log.Println("localhost Port:", localhostPort)
-
-	KAFKA_ADVERTISED_LISTENERS := fmt.Sprintf("KAFKA_ADVERTISED_LISTENERS=INTERNAL://broker:9090,EXTERNAL://localhost:%s", strconv.Itoa(localhostPortInt))
-	KAFKA_LISTENERS := "KAFKA_LISTENERS=INTERNAL://broker:9090,EXTERNAL://:9092"
-
-	log.Println("KAFKA_ADVERTISED_LISTENERS", KAFKA_ADVERTISED_LISTENERS)
-
-	resourceKafka, err := pool.RunWithOptions(&dockertest.RunOptions{
-		Repository: "confluentinc/cp-kafka",
-		Tag:        "latest",
-		NetworkID:  network.ID,
-		Hostname:   "broker",
-		PortBindings: map[dc.Port][]dc.PortBinding{
-			"29092/tcp": {{HostIP: "broker", HostPort: brokerPort}},
-			"9092/tcp":  {{HostIP: "localhost", HostPort: localhostPort}},
-		},
-		Env: []string{
-			"KAFKA_BROKER_ID=1",
-			"KAFKA_LISTENER_SECURITY_PROTOCOL_MAP=INTERNAL:PLAINTEXT,EXTERNAL:PLAINTEXT",
-			KAFKA_ADVERTISED_LISTENERS,
-			KAFKA_LISTENERS,
-			"KAFKA_ZOOKEEPER_CONNECT=zookeeper:2181",
-			"KAFKA_INTER_BROKER_LISTENER_NAME=INTERNAL",
-		},
-	})
-	if err != nil {
-		log.Panic(err)
-	}
-	log.Println("Kafka PORT:- ", resourceKafka.GetPort("9092/tcp"))
+	SetKafka()
 	defer func() {
 		if err := pool.Purge(resourceKafka); err != nil {
 			log.Printf("Could not purge resource: %s \n", err)
 		}
 	}()
-
-	// pulls an redis image, creates a container based on it and runs it
-	resourceRedis, err := pool.Run("redis", "alpine3.14", []string{"requirepass=secret"})
-	if err != nil {
-		log.Fatalf("Could not start resource: %s", err)
-	}
+	SetRedis()
 	defer func() {
 		if err := pool.Purge(resourceRedis); err != nil {
 			log.Printf("Could not purge resource: %s \n", err)
 		}
 	}()
-	// exponential backoff-retry, because the application in the container might not be ready to accept connections yet
-	redisAddress = fmt.Sprintf("localhost:%s", resourceRedis.GetPort("6379/tcp"))
-	if err := pool.Retry(func() error {
-		redisClient = redis.NewClient(&redis.Options{
-			Addr:     redisAddress,
-			Password: "",
-			DB:       0,
-		})
-
-		_, err := redisClient.Ping().Result()
-		return err
-	}); err != nil {
-		log.Fatalf("Could not connect to docker: %s", err)
+	SetJobsDB()
```

Comment:
your Set/Setup Method could return a function that could be use for defer, instead of making `resourcePostgres` global.

```go
func SetupJobsDB() func() {
  var resourcePostgres ...
 ....
  return func() {
	if err := pool.Purge(resourcePostgres); err != nil {
		log.Printf("Could not purge resource: %s \n", err)
        }
  }
}
```

Then you can do:
```suggestion
	defer SetupJobsDB()()
```

Reply:
I forgot that you need the docker resource to extract the port, in this case it would be better to just:

```go
func SetupJobsDB() *dockertest.Resource {
    var resourcePostgres ...

    return resourcePostgres
```

then:
```suggestion
	resourcePostgres := SetupJobsDB()
```